File: src/nodejs/hprose/server/HproseHttpService.js
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/

/**********************************************************\
 *                                                        *
 * HproseHttpService.js                                   *
 *                                                        *
 * HproseHttpService for Node.js.                         *
 *                                                        *
 * LastModified: Dec 6, 2012                              *
 * Author: Ma Bingyao <andot@hprfc.com>                   *
 *                                                        *
\**********************************************************/

var fs = require("fs");
var util = require("util");
var HproseService = require("./HproseService.js");

function HproseHttpService() {
    var m_crossDomain = false;
    var m_P3P = false;
    var m_get = true;
    var m_timeout = 120000;
    var m_crossDomainXmlFile = null;
    var m_crossDomainXmlContent = null;
    var m_clientAccessPolicyXmlFile = null;
    var m_clientAccessPolicyXmlContent = null;
    var m_lastModified = (new Date()).toUTCString();
    var m_etag = '"' + Math.floor(Math.random() * 2147483647).toString(16) + ":" + Math.floor(Math.random() * 2147483647).toString(16) + '"';

    var crossDomainXmlHandler = function(request, response) {
        if (request.url.toLowerCase() == "/crossdomain.xml") {
            if (request.headers["if-modified-since"] == m_lastModified &&
                request.headers["if-none-match"] == m_etag) {
                response.statusCode = 304;
            }
            else {
                response.setHeader("Last-Modified", m_lastModified);
                response.setHeader("Etag", m_etag);
                response.setHeader("Content-Type", "text/xml");
                response.setHeader("Content-Length", m_crossDomainXmlContent.length);
                response.write(m_crossDomainXmlContent);
            }
            response.end();
            return true;
        }
        return false;
    }

    var clientAccessPolicyXmlHandler = function(request, response) {
        if (request.url.toLowerCase() == "/clientaccesspolicy.xml") {
            if (request.headers["if-modified-since"] == m_lastModified &&
                request.headers["if-none-match"] == m_etag) {
                response.statusCode = 304;
            }
            else {
                response.setHeader("Last-Modified", m_lastModified);
                response.setHeader("Etag", m_etag);
                response.setHeader("Content-Type", "text/xml");
                response.setHeader("Content-Length", m_clientAccessPolicyXmlContent.length);
                response.write(m_clientAccessPolicyXmlContent);
            }
            response.end();
            return true;
        }
        return false;
    }

    HproseService.call(this);

    // protected methods
    this._sendHeader = function(request, response) {
        this.emit('sendHeader', request, response);
        response.setHeader('Content-Type', "text/plain");
        if (m_P3P) {
            response.setHeader('P3P', 
                'CP="CAO DSP COR CUR ADM DEV TAI PSA PSD IVAi IVDi ' +
                'CONi TELo OTPi OUR DELi SAMi OTRi UNRi PUBi IND PHY ONL ' +
                'UNI PUR FIN COM NAV INT DEM CNT STA POL HEA PRE GOV"');
        }
        if (m_crossDomain) {
            var origin = request.headers["origin"];
            if (origin && origin != "null") {
                response.setHeader('Access-Control-Allow-Origin', origin);
                response.setHeader('Access-Control-Allow-Credentials', 'true');  
            }
            else {
                response.setHeader('Access-Control-Allow-Origin', '*');
            }
        }
    }

    // public methods
    this.isCrossDomainEnabled = function() {
        return m_crossDomain;
    }

    this.setCrossDomainEnabled = function(enable) {
        if (enable === undefined) enable = true;
        m_crossDomain = enable;
    }

    this.isP3PEnabled = function() {
        return m_P3P;
    }

    this.setP3PEnabled = function(enable) {
        if (enable === undefined) enable = true;
        m_P3P = enable;
    }

    this.isGetEnabled = function() {
        return m_get;
    }

    this.setGetEnabled = function(enable) {
        if (enable === undefined) enable = true;
        m_get = enable;
    }

    this.getTimeout = function() {
        return m_timeout;
    }
    
    this.setTimeout = function(timeout) {
        m_timeout = timeout;
    }
    
    this.getCrossDomainXmlFile = function() {
        return m_crossDomainXmlFile;
    }

    this.setCrossDomainXmlFile = function(value) {
        m_crossDomainXmlFile = value;
        m_crossDomainXmlContent = fs.readFileSync(m_crossDomainXmlFile);
    }

    this.getCrossDomainXmlContent = function() {
        return m_crossDomainXmlContent;
    }

    this.setCrossDomainXmlContent = function(value) {
        m_crossDomainXmlFile = null;
        if (typeof(value) == "string") value = new Buffer(value);
        m_crossDomainXmlContent = value;
    }

    this.getClientAccessPolicyXmlFile = function() {
        return m_clientAccessPolicyXmlFile;
    }

    this.setClientAccessPolicyXmlFile = function(value) {
        m_clientAccessPolicyXmlFile = value;
        m_clientAccessPolicyXmlContent = fs.readFileSync(m_clientAccessPolicyXmlFile);
    }

    this.getClientAccessPolicyXmlContent = function() {
        return m_clientAccessPolicyXmlContent;
    }

    this.setClientAccessPolicyXmlContent = function(value) {
        m_clientAccessPolicyXmlFile = null;
        if (typeof(value) == "string") value = new Buffer(value);
        m_clientAccessPolicyXmlContent = value;
    }

    this.handle = function(request, response) {
        request.socket.setTimeout(m_timeout);
        var bufferList = [];
        var bufferLength = 0;
        request.on("data", function(chunk) {
            bufferList.push(chunk);
            bufferLength += chunk.length;
        });
        request.on("end", function() {
            if (m_clientAccessPolicyXmlContent != null && clientAccessPolicyXmlHandler(request, response)) return;
            if (m_crossDomainXmlContent != null && crossDomainXmlHandler(request, response)) return;
            this._sendHeader(request, response);
            response.on("end", function (data) {
                response.setHeader("Content-Length", data.length);
                response.end(data);
            });
            if ((request.method == "GET") && m_get) {
                this._doFunctionList(response);
            }
            else if (request.method == "POST") {
                var data = Buffer.concat(bufferList, bufferLength);
                this._handle(data, request, response);
            }
        }.bind(this));
    }
}

util.inherits(HproseHttpService, HproseService);

module.exports = HproseHttpService;

File: src/dotnet/src/Hprose/Server/HproseHttpService.cs
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/
/**********************************************************\
 *                                                        *
 * HproseHttpService.cs                                   *
 *                                                        *
 * hprose http service class for C#.                      *
 *                                                        *
 * LastModified: Feb 18, 2014                             *
 * Author: Ma Bingyao <andot@hprose.com>                  *
 *                                                        *
\**********************************************************/
#if !(ClientOnly || ClientProfile)
using System;
using System.IO;
using System.IO.Compression;
using System.Web;
using System.Web.SessionState;
using Hprose.Common;

namespace Hprose.Server {

    public class HproseHttpService : HproseService {

        private bool crossDomainEnabled = false;
        private bool p3pEnabled = false;
        private bool getEnabled = true;
        private bool compressionEnabled = false;
        public event SendHeaderEvent OnSendHeader = null;
        [ThreadStatic]
        private static HttpContext currentContext;

        protected override object[] FixArguments(Type[] argumentTypes, object[] arguments, int count) {
            if (argumentTypes.Length != count) {
                object[] args = new object[argumentTypes.Length];
                System.Array.Copy(arguments, args, count);
                Type argType = argumentTypes[count];
                if (argType == typeof(HttpContext)) {
                    args[count] = currentContext;
                }
                else if (argType == typeof(HttpRequest)) {
                    args[count] = currentContext.Request;
                }
                else if (argType == typeof(HttpResponse)) {
                    args[count] = currentContext.Response;
                }
                else if (argType == typeof(HttpServerUtility)) {
                    args[count] = currentContext.Server;
                }
                else if (argType == typeof(HttpApplicationState)) {
                    args[count] = currentContext.Application;
                }
                else if (argType == typeof(HttpSessionState)) {
                    args[count] = currentContext.Session;
                }
                return args;
            }
            return arguments;
        }

        public override HproseMethods GlobalMethods {
            get {
                if (gMethods == null) {
                    gMethods = new HproseHttpMethods();
                }
                return gMethods;
            }
        }

        public static HttpContext CurrentContext {
            get {
                return currentContext;
            }
        }

        public bool IsCrossDomainEnabled {
            get {
                return crossDomainEnabled;
            }
            set {
                crossDomainEnabled = value;
            }
        }

        public bool IsP3pEnabled {
            get {
                return p3pEnabled;
            }
            set {
                p3pEnabled = value;
            }
        }

        public bool IsGetEnabled {
            get {
                return getEnabled;
            }
            set {
                getEnabled = value;
            }
        }

        public bool IsCompressionEnabled {
            get {
                return compressionEnabled;
            }
            set {
                compressionEnabled = value;
            }
        }

        private Stream GetOutputStream() {
            Stream ostream = new BufferedStream(currentContext.Response.OutputStream);
            if (compressionEnabled) {
                string acceptEncoding = currentContext.Request.Headers["Accept-Encoding"];
                if (acceptEncoding != null) {
                    acceptEncoding = acceptEncoding.ToLower();
                    if (acceptEncoding.IndexOf("deflate") > -1) {
                        ostream = new DeflateStream(ostream, CompressionMode.Compress);
                    }
                    else if (acceptEncoding.IndexOf("gzip") > -1) {
                        ostream = new GZipStream(ostream, CompressionMode.Compress);
                    }
                }
            }
            return ostream;
        }

        private Stream GetInputStream() {
            Stream istream = new BufferedStream(currentContext.Request.InputStream);
            return istream;
        }

        private void SendHeader() {
            if (OnSendHeader != null) {
                OnSendHeader();
            }
            currentContext.Response.ContentType = "text/plain";
            if (p3pEnabled) {
                currentContext.Response.AddHeader("P3P", "CP=\"CAO DSP COR CUR ADM DEV TAI PSA PSD " +
                                                         "IVAi IVDi CONi TELo OTPi OUR DELi SAMi " +
                                                         "OTRi UNRi PUBi IND PHY ONL UNI PUR FIN " +
                                                         "COM NAV INT DEM CNT STA POL HEA PRE GOV\"");
            }
            if (crossDomainEnabled) {
                string origin = currentContext.Request.Headers["Origin"];
                if (origin != null && origin != "" && origin != "null") {
                    currentContext.Response.AddHeader("Access-Control-Allow-Origin", origin);
                    currentContext.Response.AddHeader("Access-Control-Allow-Credentials", "true");
                }
                else {
                    currentContext.Response.AddHeader("Access-Control-Allow-Origin", "*");
                }
            }
            if (compressionEnabled) {
                string acceptEncoding = currentContext.Request.Headers["Accept-Encoding"];
                if (acceptEncoding != null) {
                    acceptEncoding = acceptEncoding.ToLower();
                    if (acceptEncoding.IndexOf("deflate") > -1) {
                        currentContext.Response.AddHeader("Content-Encoding", "deflate");
                    }
                    else if (acceptEncoding.IndexOf("gzip") > -1) {
                        currentContext.Response.AddHeader("Content-Encoding", "gzip");
                    }
                }
            }
        }

        public void Handle() {
            Handle(HttpContext.Current, null);
        }

        public void Handle(HproseMethods methods) {
            Handle(HttpContext.Current, (HproseHttpMethods) methods);
        }

        public void Handle(HttpContext context) {
            Handle(context, null);
        }

        public void Handle(HttpContext context, HproseHttpMethods methods) {
            currentContext = context;
            try {
                SendHeader();
                string method = currentContext.Request.HttpMethod;
                Stream istream = null, ostream = null;
                if ((method == "GET") && getEnabled) {
                    ostream = GetOutputStream();
                    DoFunctionList(ostream, methods);
                }
                else if (method == "POST") {
                    istream = GetInputStream();
                    ostream = GetOutputStream();
                    Handle(istream, ostream, methods);
                }
                else {
                    currentContext.Response.StatusCode = 403;
                }
                if (istream != null) istream.Close();
                if (ostream != null) ostream.Close();
                currentContext.Response.End();
            }
            finally {
                currentContext = null;
            }
        }
    }
}
#endif